File: exercise/scala/day20/test/controllers/ReindeerSpec.scala
package controllers

import org.scalatestplus.play._
import org.scalatestplus.play.guice._
import play.api.libs.json.Json
import play.api.test.Helpers._
import play.api.test._

import java.util.UUID

class ReindeerSpec extends PlaySpec with GuiceOneAppPerTest with Injecting {
  "ReindeerController GET" should {
    "existing reindeer" in {
      val request = FakeRequest(GET, "/reindeer/40f9d24d-d3e0-4596-adc5-b4936ff84b19")
      status(route(app, request).get) mustBe OK
    }

    "NOT FOUND for a non-existing reindeer" in {
      val nonExistingReindeer = UUID.randomUUID().toString
      val request = FakeRequest(GET, s"/reindeer/$nonExistingReindeer")

      status(route(app, request).get) mustBe NOT_FOUND
    }
  }

  "ReindeerController POST" should {
    "return CONFLICT when trying to create an existing reindeer" in {
      val json = Json.obj(
        "name" -> "Petar",
        "color" -> "Purple"
      )
      val request = FakeRequest(POST, "/reindeer").withJsonBody(json)
      val result = route(app, request).get
      status(result) mustBe CONFLICT
    }
  }
}

File: exercise/ts/day20/tests/reindeer.spec.ts
import {Server} from 'http';
import {v4 as uuidv4} from 'uuid';
import {app} from '../src';
import {ReindeerColor} from '../src/types';
import {AddressInfo} from 'net';

let server: Server;
let baseUrl: string;

beforeAll((done) => {
    server = app.listen(() => {
        const {port} = server.address() as AddressInfo;
        baseUrl = `http://localhost:${port}`;
        done();
    });
});

afterAll((done) => {
    server.close(done);
});

describe('Reindeer API', () => {
    it('should get a reindeer', async () => {
        const response = await fetch(`${baseUrl}/reindeer/40f9d24d-d3e0-4596-adc5-b4936ff84b19`);
        expect(response.status).toBe(200);
    });

    it('should return not found for non-existing reindeer', async () => {
        const nonExistingReindeer = uuidv4();
        const response = await fetch(`${baseUrl}/reindeer/${nonExistingReindeer}`);
        expect(response.status).toBe(404);
    });

    it('should return conflict when trying to create an existing reindeer', async () => {
        const requestPayload = {
            name: 'Petar',
            color: ReindeerColor.Purple
        };
        const response = await fetch(`${baseUrl}/reindeer`, {
            method: 'POST',
            headers: {'Content-Type': 'application/json'},
            body: JSON.stringify(requestPayload)
        });
        expect(response.status).toBe(409);
    });
});